### PR TITLE
Hash#freeze

### PR DESCRIPTION
**Pull request by janbiedermann**
Mostly working, well pragmatically working in a acceptable way

I still need to check performance impact of using Object.freeze, maybe it has to go and using the $$frozen flag has to be sufficient, or maybe Object.freeze can be optional.

Object.freeze on the Hash instance causes lots of problems with other specs.

If in matz ruby one does:
```
a = Hash.new.freeze
a.define_singleton_method(:b) { true } # >> FrozenError
```

But that doesnt happen yet with this implementation.
Either $$frozen can be accepted globally, so that other methods can check, or Object.freeze on the instance can be made to work. Not sure yet, need to investigate the problems with the other specs further.

### CODE REVIEW

**Comment by janbiedermann** [comment]
The specs, that are not enabled yet, depend on Object#freeze or String#freeze

**Comment by janbiedermann** [comment]
Performance penalty by Object.freeze is not too bad for the use cases i tried, adds ~1.4% for example when freezing state and props during render cycle.

File: opal/corelib/error.rb [hunk 1]
```diff
@@ -224,7 +224,6 @@ class ::ZeroDivisionError   < ::StandardError; end
 class ::NameError           < ::StandardError; end
 class ::NoMethodError         < ::NameError; end
 class ::RuntimeError        < ::StandardError; end
-class ::FrozenError           < ::RuntimeError; end
```
Comment (hmdne) [inline opal/corelib/error.rb:227]: This shouldn't be done here, the error tree is of an importance for maintaining.

File: opal/corelib/error.rb [hunk 2]
```diff
@@ -224,7 +224,6 @@ class ::ZeroDivisionError   < ::StandardError; end
...
 class ::LocalJumpError      < ::StandardError; end
 class ::TypeError           < ::StandardError; end
 class ::ArgumentError       < ::StandardError; end
```
Comment (hmdne) [inline opal/corelib/error.rb:229]: eg. below is UncaughtThrowError, despite it being extended later

Comment (janbiedermann) [reply opal/corelib/error.rb]: done

**Comment by janbiedermann** [comment]
Ill open another PR with Hash#freeze, Array#freeze and Module#freeze soon, lets see how that goes

**Comment by janbiedermann** [comment]
Superseeded by #2448

**Pull request closed without merging**